File: progLang/Assignments/secondAssignment/SaveTheCat/savethecat.py
import sys
from collections import deque

class itemSymbol: # Class of cat(A) water(W) elements
    def __init__(self, line, column, symbol, time, position):
        self.line = line; self.column = column; self.symbol = symbol; self.time = time; self.position = position

def validNeighbors(line, column, symbol):
# A function that returns the valid neighbors of an element given

    neighbors = [] # Neighbors are being returned in a list of coordinates (line, column)

    if symbol != 'P':
        for ln in (line-1, line+1):
            if map[ln][column] != 'X' and map[ln][column] != symbol:
                if symbol == 'A':
                    if map[ln][column] != 'W':
                        neighbors.append([ln, column])
                        leastPossPad(ln, column)
                else:
                    neighbors.append([ln, column])
        for cl in (column-1, column+1):
            if map[line][cl] != 'X' and map[line][cl] != symbol:
                if symbol == 'A':
                    if map[line][cl] != 'W':
                        neighbors.append([line, cl])
                        leastPossPad(line, cl)
                else:
                    neighbors.append([line, cl])
    else: # Always expand pads with the order D - L- R - U = No need for comparing strings, always finding the best path
        if map[line+1][column] != 'X' and map[line+1][column] != 'P':
            neighbors.append([line+1, column])
        if map[line][column-1] != 'X' and map[line][column-1] != 'P':
            neighbors.append([line, column-1])
        if map[line][column+1] != 'X' and map[line][column+1] != 'P':
            neighbors.append([line, column+1])
        if map[line-1][column] != 'X' and map[line-1][column] != 'P':
            neighbors.append([line-1, column])

    return neighbors

def leastPossPad(line, column):
# A function that returns the upper and righter possible position of Arjumand possible at each specific time

    global leastl; global leastc

    if line < leastl: # Prioritazing line over column in the best possible position
        leastl = line; leastc = column
    elif line == leastl:
        if column < leastc:
            leastl = line; leastc = column

map = []
queue = deque()

# Map padding
with open(sys.argv[1]) as fileobj:
    for line in fileobj:
        # Line reading and border of X's added to each line
        map.append(['X'] + list(line.rstrip()) + ['X'])
fileobj.close()

# Lines(N) and columns(M) calculation
N = len(map)
M = len(map[0]) - 2

# Border creation of X's added to first and last line of map
map.insert(0, ['X' for i in range (M+2)])
map.insert(N+1, ['X' for i in range (M+2)])

time = 0
# Enqueing water and cat elements
for i in range (1, N+1):
    for j in range (1, M+1):
        ch = map[i][j]
        if ch is 'W' or ch is 'A':
            if ch is 'A':
                leastl = i; leastc = j # Starting pad of Arjumand
                start_line = i; start_column = j
            ch = itemSymbol(i, j, ch, time, "")
            queue.append(ch)

arjumand = 0 # Initialization of time variables
global_time = 0

empty = False # Queue flag
element = queue[0] # Getting first element without popping it - covers the zero flood case ;)

while not empty: # Floodfilling W's and A elements

    while element.time == global_time:

        temp_element = validNeighbors(element.line, element.column, element.symbol)

        for item in temp_element:
            temp_line = item[0]; temp_column = item[1]
            if element.symbol == 'W':
                if map[temp_line][temp_column] == 'A':
                    if global_time > arjumand: # Keeping coordinates of best possible time
                        arjumand = global_time
                        lpath = temp_line; cpath = temp_column
                    elif global_time == arjumand: # Minimizing string sequence
                        if temp_line <= lpath:
                            if temp_column <= cpath:
                                lpath = temp_line; cpath = temp_column
            ch = itemSymbol(temp_line, temp_column, element.symbol, time+1, "")
            map[temp_line][temp_column] = element.symbol
            queue.append(ch)

        if queue: # If queue is not empty go to next element, else break time - loop
            element = queue.popleft()
            time = element.time
        else:
            empty = True
            global_time = -1
    global_time += 1

ch = itemSymbol(start_line, start_column, 'P', 0, "") # Initializing Arjumand for path finding
map[start_line][start_column] = 'P'
queue.append(ch)

flag = 0 # Path is found
global_time = 0

empty = False
element = queue[0]

while flag != 1:

    while element.time == global_time:

        temp_element = validNeighbors(element.line, element.column, element.symbol)

        for item in temp_element:
            temp_line = item[0]; temp_column = item[1]
            if temp_line > element.line: # Prioritazing: D - L - R - U
                pos = element.position + "D"
            if temp_column < element.column:
                pos = element.position + "L"
            if temp_column > element.column:
                pos = element.position + "R"
            if temp_line < element.line:
                pos = element.position + "U"
            ch = itemSymbol(temp_line, temp_column, element.symbol, time+1, pos)
            map[temp_line][temp_column] = element.symbol
            queue.append(ch)
            if arjumand != 0:
                if temp_line == lpath and temp_column == cpath:
                    path = pos
                    flag = 1
            else:
                if temp_line == leastl and temp_column == leastc:
                    road = pos
                    flag = 1

        if queue:
            element = queue.popleft()
            time = element.time
        else:
            empty = True
            global_time = -1
    global_time += 1

if arjumand == 0:
    print("infinity")
    if road == "":
        print("stay")
    else:
        print(road)
else:
    print(arjumand)
    if path == "":
        print("stay")
    else:
        print(path)

File: 1-programming-languages/Assignments/1/SaveTheCat/savethecat.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct node{
    int line; int column; char symbol; int time; char* position;
    struct node *next;
};

struct node *front = NULL;
struct node *rear = NULL;

void enqueueElem(int x, int y, char ch, int time){
/* Enqueing wihtout keeping path of elements */

    struct node *nptr = malloc(sizeof(struct node));
    nptr->line = x; nptr->column = y; nptr->symbol = ch; nptr->time = time; nptr->next = NULL;
    if (rear == NULL){
        front = nptr;
        rear = nptr;
    }
    else{
        rear->next = nptr;
        rear = rear->next;
    }
}

void enqueueCat(int x, int y, char ch, int time, char* pos){
/* Enqueing and keeping path of cat elemets */

    struct node *nptr = malloc(sizeof(struct node));
    nptr->line = x; nptr->column = y; nptr->symbol = ch; nptr->time = time; nptr->position = pos; nptr->next = NULL;
    if (rear == NULL){
        front = nptr;
        rear = nptr;
    }
    else{
        rear->next = nptr;
        rear = rear->next;
    }
}

void dequeue(){
/* Standard dequeue function */

    if (front == NULL){
        printf("\n\n Queue is empty \n");
    }
    else{
        struct node *temp;
        temp = front;
        front = front->next;
        free(temp);
    }
}

struct Tuple{ // Returning multiple values from function - https://bit.ly/2W0Ym31
    int arjumand;
    int line;
    int column;
};

struct Follow{
    char* path;
    char* road;
    int flag;
};

char* stringConcat(char* position, char* destination){
/* A function that concatenates two strings given and returns the result - https://bit.ly/2zVGJZS */

    char* str = (char*) malloc(1 + strlen(position)+ strlen(destination));
    strcpy(str, position);
    strcat(str, destination);
    return str;
}

struct Tuple road_coordFinding(int line, int column, int leastl, int leastc){
/* A function that returns the coordinates of the best possible position of Arjumand when she is safe */

    struct Tuple value;

    if (line < leastl) { // Prioritazing line over column in best possible pad
        leastl = line;
        leastc = column;
    }
    else if (line == leastl){
        if (column < leastc){
            leastl = line;
            leastc = column;
        }
    }
    value.line = leastl; value.column = leastc;
    return value;
}

struct Tuple path_coordFinding(int line, int column, int lpath, int cpath, int global_time, int arjumand){
/* A function that returns coordinates of the best possible location of Arjumand when she needs to be saved*/

    struct Tuple result;

    if (global_time - 1 == arjumand){
        if (line < lpath){ // Prioritazing line over column in best possible pad
            lpath = line;
            cpath = column;
        }
        else if (line == lpath){
            if (column < cpath){
                lpath = line;
                cpath = column;
            }
        }
    }
    else if (global_time > arjumand){
        arjumand = global_time - 1;
        lpath = line;
        cpath = column;
    }
    result.arjumand = arjumand; result.line = lpath; result.column = cpath;
    return result;
}

struct Follow pathFinding(int arjumand, int line, int column, int lpath, int cpath, int leastl, int leastc, char* str, char* path, char* road, int flag){
/* A function that returns the path Arjumand ought to follow */

    struct Follow final;

    if (arjumand != 0){
        if ((line == lpath) && (column == cpath)){
            path = str;
            flag = 1;
        }
    }
    if (arjumand == 0){
        if ((line == leastl) && (column == leastc)){
            road = str;
            flag = 1;
        }
    }
    final.path = path; final.road = road; final.flag = flag;
    return final;
}

int main(int argc, char *argv[]) {

    char ch;
    int i, j, N = 0, M = 0, flag = 0;

    FILE *fp = fopen(argv[1], "r");

    /* Lines and columns calculation */
    while (!feof(fp)){
        ch = fgetc(fp);
        if(ch == '\n'){
            N++;
            flag = 1;
        }
        if (flag == 0){
            M++;
        }
    }
    fclose (fp);
    char array[N+2][M+2];

    int leastl = 0, leastc = 0; // Keeping coordinates of the best possible cat cell on the map
    int start_line = 0, start_column = 0; // Starting coordinates of Arjumand

    /* File reading and map padding */
    fp = fopen(argv[1], "r");
    for (i = 1; i <= N+1; i++){
        for (j = 1; j <= M+1; j++){
            ch = fgetc(fp);
            array[i][j] = ch;
            if (ch == 'W'){
                int time = 1;
                enqueueElem(i, j, ch, time);
            }
            if (ch == 'A'){
                int time = 1;
                enqueueElem(i, j, ch, time);
                leastl = i; leastc = j; start_line = i; start_column = j;
            }
        }
    }

    /* Border creation of X's */
    for (i = 0; i < M+2; i++){
        array[0][i] = 'X'; array[N+1][i] = 'X';
    }
    for (i = 0; i < N+2; i++){
        array[i][0] = 'X'; array[i][M+1] = 'X';
    }

    int arjumand, lpath, cpath;
    int global_time = 0, time = front->time;

    struct Tuple value; // Keeping coordinates of the best possible position of Arjumand
    struct Tuple result; // Keeping coordinates of the path Arjumand and the longest possible time of rescue
    value.line = leastl; value.column = leastc;
    result.line = N; result.column = M; result.arjumand = 0;

    /* Floodfilling A's and W's */
    while (front != NULL){
        while (global_time == time){
            int line = front->line, column = front->column;
            char item = front->symbol;
            /* Cross elements for the tested item */
            char item_east = array[line][column+1], item_west = array[line][column-1];
            char item_north = array[line+1][column], item_south = array[line-1][column];
            if (item == 'A'){
                if ((item_east != 'W') && (item_east != 'X') && (item_east != 'A')){
                    array[line][column+1] = item;
                    enqueueElem(line, column+1, item, time+1);
                    value = road_coordFinding(line, column+1, value.line, value.column);
                }
                if ((item_north != 'W') && (item_north != 'X') && (item_north != 'A')){
                    array[line+1][column] = item;
                    enqueueElem(line+1, column, item, time+1);
                    value = road_coordFinding(line+1, column, value.line, value.column);
                }
                if ((item_west != 'W') && (item_west != 'X') && (item_west != 'A')){
                    array[line][column-1] = item;
                    enqueueElem(line, column-1, item, time+1);
                    value = road_coordFinding(line, column-1, value.line, value.column);
                }
                if ((item_south != 'W') && (item_south != 'X') && (item_south != 'A')){
                    array[line-1][column] = item;
                    enqueueElem(line-1, column, item, time+1);
                    value = road_coordFinding(line-1, column, value.line, value.column);
                }
            }
            if (item == 'W'){
                if ((item_west != 'W') && (item_west != 'X') && (item_west != 'A')){
                    array[line][column-1] = item;
                    enqueueElem(line, column-1, item, time+1);
                }
                else if (item_west == 'A'){
                    array[line][column-1] = item;
                    enqueueElem(line, column-1, item, time+1);
                    result = path_coordFinding(line, column-1, result.line, result.column, global_time, result.arjumand);
                }
                if ((item_east != 'W') && (item_east != 'X') && (item_east != 'A')){
                    array[line][column+1] = item;
                    enqueueElem(line, column+1, item, time+1);
                }
                else if (item_east == 'A'){
                    array[line][column+1] = item;
                    enqueueElem(line, column+1, item, time+1);
                    result = path_coordFinding(line, column+1, result.line, result.column, global_time, result.arjumand);
                }
                if ((item_north != 'W') && (item_north != 'X') && (item_north != 'A')){
                    array[line+1][column] = item;
                    enqueueElem(line+1, column, item, time+1);
                }
                else if (item_north == 'A'){
                    array[line+1][column] = item;
                    enqueueElem(line+1, column, item, time+1);
                    result = path_coordFinding(line+1, column, result.line, result.column, global_time, result.arjumand);
                }
                if ((item_south != 'W') && (item_south != 'X') && (item_south != 'A')){
                    array[line-1][column] = item;
                    enqueueElem(line-1, column, item, time+1);
                }
                else if (item_south == 'A'){
                    array[line-1][column] = item;
                    enqueueElem(line-1, column, item, time+1);
                    result = path_coordFinding(line-1, column, result.line, result.column, global_time, result.arjumand);
                }
            }
            dequeue();

            /* If queue is not empty, go to the next element */
            if (front != NULL){
                time = front->time;
            }
            else{
                global_time = -1;
            }
        }
        global_time++;
    }

    leastl = value.line; leastc = value.column;
    arjumand = result.arjumand; lpath = result.line; cpath = result.column;

    front = NULL, rear = NULL; // Reinitialization of queue with the first cat element

    /* Enqueuing starting position of Arjumand */
    array[start_line][start_column] = 'P', enqueueCat(start_line, start_column, 'P', 1, ""); // P for path ;)

    char* str; // Temp variable for string concatenation

    struct Follow final;
    final.path = ""; final.road = ""; final.flag = 0; // Final paths

    /* Floodfilling A's for path finding  */
    while ((front != NULL) && (final.flag != 1)){
        while (global_time == time){
            int line = front->line, column = front->column;
            char item = front->symbol;
            char *pos = front->position;
            /* Cross elements for the tested item */
            char item_east = array[line][column + 1], item_west = array[line][column - 1];
            char item_north = array[line + 1][column], item_south = array[line - 1][column];
            if (item == 'P'){
                if ((item_north != 'X') && (item_north != 'P')){
                    array[line+1][column] = item;
                    str = stringConcat(pos, "D");
                    enqueueCat(line+1, column, item, time+1, str);
                    final = pathFinding(arjumand, line+1, column, lpath, cpath, leastl, leastc, str, final.path, final.road, final.flag);
                }
                if ((item_west != 'X') && (item_west != 'P')){
                    array[line][column-1] = item;
                    str = stringConcat(pos, "L");
                    enqueueCat(line, column-1, item, time+1, str);
                    final = pathFinding(arjumand, line, column-1, lpath, cpath, leastl, leastc, str, final.path, final.road, final.flag);
                }
                if ((item_east != 'X') && (item_east != 'P')){
                    array[line][column+1] = item;
                    str = stringConcat(pos, "R");
                    enqueueCat(line, column+1, item, time+1, str);
                    final = pathFinding(arjumand, line, column+1, lpath, cpath, leastl, leastc, str, final.path, final.road, final.flag);
                }
                if ((item_south != 'X') && (item_south != 'P')){
                    array[line-1][column] = item;
                    str = stringConcat(pos, "U");
                    enqueueCat(line-1, column, item, time+1, str);
                    final = pathFinding(arjumand, line-1, column, lpath, cpath, leastl, leastc, str, final.path, final.road, final.flag);
                }
            }
            dequeue();

            if (front != NULL){ // If queue is not empty go to the next element
                time = front->time;
            }
            else{
                global_time = -1;
            }
        }
        global_time++;
    }

    /* In case Arjumand is safe */
    if (arjumand == 0){
        printf("infinity\n");
        if (strcmp(final.road, "") == 0){
            printf("stay");
        }
        else{
            printf("%s", final.road);
        }
    }
    /* In case Arjumand ought to be saved */
    else{
        printf("%d\n", arjumand);
        if (strcmp(final.path, "") == 0){
            printf("stay");
        }
        else{
            printf("%s", final.path);
        }
    }

    return 0;
}